pyomo/contrib/solver/gurobi_direct.py

```diff
@@ -0,0 +1,419 @@
+            )
+            for v in repn.columns
+        ]
+        sense_type = '=<>'  # Note: ordering matches 0, 1, -1
+        sense = [sense_type[r[1]] for r in repn.rows]
+        timer.stop('prepare_matrices')
+
+        ostreams = [io.StringIO()] + config.tee
+
+        try:
+            orig_cwd = os.getcwd()
+            if config.working_dir:
+                os.chdir(config.working_dir)
+            with TeeStream(*ostreams) as t, capture_output(t.STDOUT, capture_fd=False):
+                gurobi_model = gurobipy.Model()
+
+                timer.start('transfer_model')
+                x = gurobi_model.addMVar(
+                    len(repn.columns),
+                    lb=lb,
+                    ub=ub,
+                    obj=repn.c.todense()[0] if repn.c.shape[0] else 0,
+                    vtype=vtype,
+                )
+                A = gurobi_model.addMConstr(repn.A, x, sense, repn.rhs)
+                if repn.c.shape[0]:
+                    gurobi_model.setAttr('ObjCon', repn.c_offset[0])
+                    gurobi_model.setAttr('ModelSense', int(repn.objectives[0].sense))
+                # gurobi_model.update()
+                timer.stop('transfer_model')
+
+                options = config.solver_options
+
+                gurobi_model.setParam('LogToConsole', 1)
+
+                if config.threads is not None:
+                    gurobi_model.setParam('Threads', config.threads)
+                if config.time_limit is not None:
+                    gurobi_model.setParam('TimeLimit', config.time_limit)
+                if config.rel_gap is not None:
+                    gurobi_model.setParam('MIPGap', config.rel_gap)
+                if config.abs_gap is not None:
+                    gurobi_model.setParam('MIPGapAbs', config.abs_gap)
+
+                if config.use_mipstart:
+                    raise MouseTrap("MIPSTART not yet supported")
+
+                for key, option in options.items():
+                    gurobi_model.setParam(key, option)
+
+                timer.start('optimize')
+                gurobi_model.optimize()
+                timer.stop('optimize')
+        finally:
+            os.chdir(orig_cwd)
+
+        res = self._postsolve(
+            timer,
+            config,
+            GurobiDirectSolutionLoader(
+                gurobi_model, A, x, repn.rows, repn.columns, repn.objectives
+            ),
+        )
+        res.solver_configuration = config
+        res.solver_name = 'Gurobi'
+        res.solver_version = self.version()
+        res.solver_log = ostreams[0].getvalue()
+
+        end_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        res.timing_info.start_timestamp = start_timestamp
+        res.timing_info.wall_time = (end_timestamp - start_timestamp).total_seconds()
+        res.timing_info.timer = timer
+        return res
+
+    def _postsolve(self, timer: HierarchicalTimer, config, loader):
+        grb_model = loader._grb_model
```

Comment:
Something about this is making me pause. It's not related to this PR, but should `_grb_model` be "private"?

Reply:
In this case, yes: the underlying Gurobi model is never meant to be interacted with directly by the user.

(I also copied this from the persistent version in `contrib/solver`)